application-task-manager-default/src/main/java/com/xwiki/taskmanager/internal/TaskMacroUpdateEventListener.java

```diff
@@ -0,0 +1,219 @@
+    {
+        super(TaskMacroUpdateEventListener.class.getName(), Arrays.asList(new DocumentUpdatingEvent(),
+            new DocumentCreatingEvent(), new DocumentDeletingEvent()));
+    }
+
+    @Override
+    protected void processEvent(XWikiDocument document, XWikiContext context, Event event)
+    {
+        if (event instanceof DocumentDeletingEvent) {
+            try {
+                taskManager.deleteTasksByOwner(document.getDocumentReference());
+            } catch (TaskException e) {
+                logger.warn(e.getMessage(), e);
+            }
+            return;
+        }
+        if (context.get(TASK_UPDATE_FLAG) != null) {
+            return;
+        }
+        XDOM documentContent = document.getXDOM();
+
+        List<Task> tasks = this.taskXDOMProcessor.extract(documentContent, document.getDocumentReference());
+
+        String previousVersion = document.getPreviousVersion();
+        List<Task> previousDocTasks = new ArrayList<>();
+
+        if (previousVersion != null) {
+            try {
+                XWikiDocument previousVersionDoc = revisionProvider.getRevision(document, previousVersion);
+                XDOM previousContent = previousVersionDoc.getXDOM();
+                previousDocTasks = this.taskXDOMProcessor.extract(previousContent, document.getDocumentReference());
+                List<DocumentReference> currentTasksIds =
+                    tasks.stream().map(Task::getReference).collect(Collectors.toList());
+                previousDocTasks.removeIf(task -> currentTasksIds.contains(task.getReference()));
+            } catch (XWikiException ignored) {
+                logger.warn(
+                    "Could not check for the possibly removed Task Macros and delete their associated Task Pages.");
+            }
+        }
+        if (tasks.size() > 0 || previousDocTasks.size() > 0) {
+            // TaskExtractor will add new IDs to the tasks if they don't have one, so we need to update the content.
+            try {
+                document.setContent(documentContent);
+            } catch (XWikiException e) {
+                logger.error("Could not update the content of the document!");
+            }
+            context.put(TASK_UPDATE_FLAG, true);
+            deleteTaskPages(document, context, previousDocTasks);
+            createOrUpdateTaskPages(document, context, tasks);
+            context.put(TASK_UPDATE_FLAG, null);
+        }
+    }
+
+    private void deleteTaskPages(XWikiDocument document, XWikiContext context, List<Task> previousDocTasks)
+    {
+        for (Task previousDocTask : previousDocTasks) {
+            try {
+                XWikiDocument taskDoc = context.getWiki().getDocument(previousDocTask.getReference(), context);
+                BaseObject taskObj = taskDoc.getXObject(TASK_CLASS_REFERENCE);
+                if (!document.getDocumentReference()
+                    .equals(resolver.resolve(taskObj.getLargeStringValue(Task.OWNER))))
```

Comment:
```suggestion
                    .equals(resolver.resolve(taskObj.getLargeStringValue(Task.OWNER), task.getReference())))
```

The owner reference should be resolved relative to to task reference.